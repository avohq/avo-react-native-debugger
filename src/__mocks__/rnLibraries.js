jest.mock('react-native/Libraries/Settings/Settings.ios.js', () => ({
  get: jest.fn(() => null),
  set: jest.fn()
}));

jest.mock('react-native/Libraries/Utilities/BackHandler.ios.js', () => ({
  addEventListener: jest.fn(),
  removeEventListener: jest.fn(),
}));
