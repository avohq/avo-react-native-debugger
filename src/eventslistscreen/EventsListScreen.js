import React from 'react';
import {
  View,
  FlatList,
  BackHandler,
  Image,
  TouchableOpacity
} from 'react-native';
import {Component} from 'react';
import RootSiblings from 'react-native-root-siblings';
import PropTypes from 'prop-types';
import EventsListItem from './EventListItem';
import {images} from '../images';
import {topOffset, eventsHaveErrors} from '../utils';
import {styles} from './EventsListScreenStyles';
import {baseStyles} from '../BaseStyles';
import AvoDebugger from '../AvoDebugger';

export default class EventsListScreen extends Component {
  static logScreen = null;

  static isVisible = () => {
    return EventsListScreen.logScreen != null;
  }

  static updateDebuggerLogScreen = () => {
    if (EventsListScreen.logScreen) {
      EventsListScreen.logScreen.destroy();
      EventsListScreen.logScreen = new RootSiblings(
        <EventsListScreen items={AvoDebugger.items} />
      );
    }
  }

  static toggleDebuggerLogScreen = () => {
    if (EventsListScreen.logScreen == null) {
      EventsListScreen.logScreen = new RootSiblings(
        <EventsListScreen items={AvoDebugger.items} />
      );
    } else {
      EventsListScreen.logScreen.destroy();
      EventsListScreen.logScreen = null;
    }
  };

  componentWillMount() {
    BackHandler.addEventListener(
      'hardwareBackPress',
      this.handleBackButtonClick
    );
  }

  componentWillUnmount() {
    BackHandler.removeEventListener(
      'hardwareBackPress',
      this.handleBackButtonClick
    );
  }

  handleBackButtonClick() {
    EventsListScreen.toggleDebuggerLogScreen();
    return true;
  }

  listRow = ({item}) => (
    <EventsListItem
      item={item}
      expended={this.props.items[0] === item || eventsHaveErrors([item])}
    />
  );

  render() {
    return (
      <View style={[styles.screenContainer, {marginTop: topOffset()}]}>
        <View style={styles.avoLabelContainer}>
          <Image source={images.avo_logo} />
          <TouchableOpacity
            style={styles.closeButtonTouchableArea}
            onPress={() => {
              EventsListScreen.toggleDebuggerLogScreen();
            }}
          >
            <Image style={styles.closeButton} source={images.close_button} />
          </TouchableOpacity>
        </View>
        <View style={baseStyles.divider} />

        <FlatList
          style={styles.list}
          data={this.props.items.sort((a, b) => { return b.timestamp - a.timestamp })}
          renderItem={this.listRow}
        />
      </View>
    );
  }
}

EventsListScreen.propTypes = {
  items: PropTypes.arrayOf(
    PropTypes.exact({
      key: PropTypes.string.isRequired,
      id: PropTypes.string.isRequired,
      timestamp: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      messages: PropTypes.arrayOf(
        PropTypes.exact({
          tag: PropTypes.string,
          propertyId: PropTypes.string.isRequired,
          message: PropTypes.string.isRequired,
          allowedTypes: PropTypes.arrayOf(PropTypes.string),
          providedType: PropTypes.string
        })
      ),
      eventProps: PropTypes.arrayOf(
        PropTypes.exact({
          id: PropTypes.string.isRequired,
          name: PropTypes.string.isRequired,
          value: PropTypes.any
        })
      ),
      userProps: PropTypes.arrayOf(
        PropTypes.exact({
          id: PropTypes.string.isRequired,
          name: PropTypes.string.isRequired,
          value: PropTypes.any
        })
      )
    })
  )
};
