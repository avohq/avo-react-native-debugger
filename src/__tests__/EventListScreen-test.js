import React from "react";
import * as ReactNative from "react-native";
import renderer from "react-test-renderer";

import "../__mocks__/rnLibraries";

import EventsListScreen from "../eventslistscreen/EventsListScreen";
import { testEventItem } from "./testUtils";

describe("EventListScreen", () => {
  const items = [
    {
      ...testEventItem,
      key: "key-0",
      id: "id-0",
      timestamp: +new Date(2020, 9, 15, 12, 4, 36)
    },
    {
      ...testEventItem,
      key: "key-1",
      id: "id-1",
      timestamp: +new Date(2020, 9, 15, 12, 4, 35)
    },
    {
      ...testEventItem,
      key: "key-2",
      id: "id-2",
      timestamp: +new Date(2020, 9, 16)
    }
  ];

  test("Renders correctly", () => {
    const tree = renderer.create(<EventsListScreen items={items} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("Toggle opens and closes event list", () => {
    expect(EventsListScreen.logScreen).toBeNull();

    EventsListScreen.toggleDebuggerLogScreen();

    expect(EventsListScreen.logScreen).not.toBeNull();

    EventsListScreen.toggleDebuggerLogScreen();

    expect(EventsListScreen.logScreen).toBeNull();
  });

  test("updateDebuggerLogScreen() re-renders the list view", () => {
    EventsListScreen.toggleDebuggerLogScreen();
    expect(EventsListScreen.logScreen).not.toBeNull();

    // TODO: maybe update() can be used instead of deastroy() + new?
    const destroySpy = jest.spyOn(EventsListScreen.logScreen, "destroy");

    EventsListScreen.updateDebuggerLogScreen();

    expect(destroySpy).toHaveBeenCalled();
    expect(EventsListScreen.logScreen).not.toBeNull();
  });

  test("Back handler is binded on mount and removed on destroy", () => {
    const listScreen = renderer.create(<EventsListScreen items={items} />);
    expect(ReactNative.BackHandler.addEventListener).toHaveBeenCalled();

    listScreen.unmount();
    expect(ReactNative.BackHandler.removeEventListener).toHaveBeenCalled();
  });

  test("Back event toggles list view", () => {
    const toggleSpy = jest.spyOn(EventsListScreen, "toggleDebuggerLogScreen");

    const component = renderer.create(<EventsListScreen items={items} />);
    const componentInstance = component.getInstance();

    componentInstance.handleBackButtonClick();

    expect(toggleSpy).toHaveBeenCalled();
  });

  test("Events are sorted descending by time", () => {
    const component = renderer.create(<EventsListScreen items={items} />).root;

    const sortedItems = items.sort((a, b) => b.timestamp - a.timestamp);

    expect(component.findByType(ReactNative.FlatList).props.data).toBe(
      sortedItems
    );
  });
});
