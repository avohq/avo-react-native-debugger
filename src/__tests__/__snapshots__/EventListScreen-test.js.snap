// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventListScreen Renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "bottom": 0,
        "flex": 1,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      Object {
        "marginTop": undefined,
      },
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "height": 48,
        "justifyContent": "center",
      }
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../src/img/avo_logo.png",
        }
      }
    />
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "height": 48,
          "opacity": 1,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 48,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../src/img/close_button.png",
          }
        }
        style={
          Object {
            "height": 10,
            "marginLeft": 19,
            "marginTop": 19,
            "width": 10,
          }
        }
      />
    </View>
  </View>
  <View
    style={
      Object {
        "alignSelf": "stretch",
        "backgroundColor": "#E3E3E3",
        "height": 1,
      }
    }
  />
  <RCTScrollView
    data={
      Array [
        Object {
          "eventProps": Array [
            Object {
              "id": "id0",
              "name": "prop 0 name",
              "value": "val 0",
            },
            Object {
              "id": "id1",
              "name": "prop 1 name",
              "value": "val 1",
            },
          ],
          "id": "id-2",
          "key": "key-2",
          "messages": Array [
            Object {
              "message": "error in prop 0 with id0",
              "propertyId": "id0",
            },
          ],
          "name": "test-name",
          "timestamp": 1602806400000,
          "userProps": Array [],
        },
        Object {
          "eventProps": Array [
            Object {
              "id": "id0",
              "name": "prop 0 name",
              "value": "val 0",
            },
            Object {
              "id": "id1",
              "name": "prop 1 name",
              "value": "val 1",
            },
          ],
          "id": "id-0",
          "key": "key-0",
          "messages": Array [
            Object {
              "message": "error in prop 0 with id0",
              "propertyId": "id0",
            },
          ],
          "name": "test-name",
          "timestamp": 1602763476000,
          "userProps": Array [],
        },
        Object {
          "eventProps": Array [
            Object {
              "id": "id0",
              "name": "prop 0 name",
              "value": "val 0",
            },
            Object {
              "id": "id1",
              "name": "prop 1 name",
              "value": "val 1",
            },
          ],
          "id": "id-1",
          "key": "key-1",
          "messages": Array [
            Object {
              "message": "error in prop 0 with id0",
              "propertyId": "id0",
            },
          ],
          "name": "test-name",
          "timestamp": 1602763475000,
          "userProps": Array [],
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "backgroundColor": "#FFFFFF",
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flex": 1,
              },
              Object {
                "backgroundColor": "#FFFFFF",
              },
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "height": 52,
                "opacity": 1,
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../src/img/red_warning.png",
                }
              }
              style={
                Object {
                  "height": 16,
                  "margin": 16,
                  "resizeMode": "contain",
                  "width": 16,
                }
              }
            />
            <View
              style={
                Object {
                  "flex": 1,
                  "marginLeft": 8,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontSize": 8,
                      "fontWeight": "bold",
                      "textAlignVertical": "center",
                    },
                    Object {
                      "color": "#666666",
                    },
                  ]
                }
              >
                00:00:00.000
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "fontSize": 16,
                      "fontWeight": "bold",
                      "textAlignVertical": "center",
                    },
                    Object {
                      "color": "#D94553",
                    },
                  ]
                }
              >
                test-name
              </Text>
            </View>
            <Image
              source={
                Object {
                  "testUri": "../../../src/img/collapse_arrow.png",
                }
              }
              style={
                Object {
                  "height": 6,
                  "margin": 19,
                  "width": 10,
                }
              }
            />
          </View>
          <View
            style={
              Object {
                "alignSelf": "stretch",
                "marginEnd": 16,
                "marginStart": 16,
              }
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "column",
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "height": 32,
                    "justifyContent": "space-between",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "fontSize": 12,
                        "fontWeight": "bold",
                      },
                      Object {
                        "color": "#D94553",
                      },
                    ]
                  }
                >
                  prop 0 name
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "fontSize": 12,
                      },
                      Object {
                        "color": "#D94553",
                      },
                    ]
                  }
                >
                  val 0
                </Text>
              </View>
              <Text
                style={
                  Object {
                    "color": "#D94553",
                    "fontSize": 10,
                    "marginBottom": 16,
                  }
                }
              >
                error in prop 0 with id0
              </Text>
            </View>
            <View
              style={
                Object {
                  "alignSelf": "stretch",
                  "backgroundColor": "#E3E3E3",
                  "height": 1,
                }
              }
            />
            <View
              style={
                Object {
                  "flexDirection": "column",
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "height": 32,
                    "justifyContent": "space-between",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "fontSize": 12,
                        "fontWeight": "bold",
                      },
                      Object {
                        "color": "#333333",
                      },
                    ]
                  }
                >
                  prop 1 name
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "fontSize": 12,
                      },
                      Object {
                        "color": "#333333",
                      },
                    ]
                  }
                >
                  val 1
                </Text>
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "alignSelf": "stretch",
                "backgroundColor": "#E3E3E3",
                "height": 1,
              }
            }
          />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flex": 1,
              },
              Object {
                "backgroundColor": "#FFFFFF",
              },
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "height": 52,
                "opacity": 1,
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../src/img/red_warning.png",
                }
              }
              style={
                Object {
                  "height": 16,
                  "margin": 16,
                  "resizeMode": "contain",
                  "width": 16,
                }
              }
            />
            <View
              style={
                Object {
                  "flex": 1,
                  "marginLeft": 8,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontSize": 8,
                      "fontWeight": "bold",
                      "textAlignVertical": "center",
                    },
                    Object {
                      "color": "#666666",
                    },
                  ]
                }
              >
                12:04:36.000
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "fontSize": 16,
                      "fontWeight": "bold",
                      "textAlignVertical": "center",
                    },
                    Object {
                      "color": "#D94553",
                    },
                  ]
                }
              >
                test-name
              </Text>
            </View>
            <Image
              source={
                Object {
                  "testUri": "../../../src/img/collapse_arrow.png",
                }
              }
              style={
                Object {
                  "height": 6,
                  "margin": 19,
                  "width": 10,
                }
              }
            />
          </View>
          <View
            style={
              Object {
                "alignSelf": "stretch",
                "marginEnd": 16,
                "marginStart": 16,
              }
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "column",
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "height": 32,
                    "justifyContent": "space-between",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "fontSize": 12,
                        "fontWeight": "bold",
                      },
                      Object {
                        "color": "#D94553",
                      },
                    ]
                  }
                >
                  prop 0 name
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "fontSize": 12,
                      },
                      Object {
                        "color": "#D94553",
                      },
                    ]
                  }
                >
                  val 0
                </Text>
              </View>
              <Text
                style={
                  Object {
                    "color": "#D94553",
                    "fontSize": 10,
                    "marginBottom": 16,
                  }
                }
              >
                error in prop 0 with id0
              </Text>
            </View>
            <View
              style={
                Object {
                  "alignSelf": "stretch",
                  "backgroundColor": "#E3E3E3",
                  "height": 1,
                }
              }
            />
            <View
              style={
                Object {
                  "flexDirection": "column",
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "height": 32,
                    "justifyContent": "space-between",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "fontSize": 12,
                        "fontWeight": "bold",
                      },
                      Object {
                        "color": "#333333",
                      },
                    ]
                  }
                >
                  prop 1 name
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "fontSize": 12,
                      },
                      Object {
                        "color": "#333333",
                      },
                    ]
                  }
                >
                  val 1
                </Text>
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "alignSelf": "stretch",
                "backgroundColor": "#E3E3E3",
                "height": 1,
              }
            }
          />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flex": 1,
              },
              Object {
                "backgroundColor": "#FFFFFF",
              },
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "height": 52,
                "opacity": 1,
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../src/img/red_warning.png",
                }
              }
              style={
                Object {
                  "height": 16,
                  "margin": 16,
                  "resizeMode": "contain",
                  "width": 16,
                }
              }
            />
            <View
              style={
                Object {
                  "flex": 1,
                  "marginLeft": 8,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontSize": 8,
                      "fontWeight": "bold",
                      "textAlignVertical": "center",
                    },
                    Object {
                      "color": "#666666",
                    },
                  ]
                }
              >
                12:04:35.000
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "fontSize": 16,
                      "fontWeight": "bold",
                      "textAlignVertical": "center",
                    },
                    Object {
                      "color": "#D94553",
                    },
                  ]
                }
              >
                test-name
              </Text>
            </View>
            <Image
              source={
                Object {
                  "testUri": "../../../src/img/collapse_arrow.png",
                }
              }
              style={
                Object {
                  "height": 6,
                  "margin": 19,
                  "width": 10,
                }
              }
            />
          </View>
          <View
            style={
              Object {
                "alignSelf": "stretch",
                "marginEnd": 16,
                "marginStart": 16,
              }
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "column",
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "height": 32,
                    "justifyContent": "space-between",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "fontSize": 12,
                        "fontWeight": "bold",
                      },
                      Object {
                        "color": "#D94553",
                      },
                    ]
                  }
                >
                  prop 0 name
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "fontSize": 12,
                      },
                      Object {
                        "color": "#D94553",
                      },
                    ]
                  }
                >
                  val 0
                </Text>
              </View>
              <Text
                style={
                  Object {
                    "color": "#D94553",
                    "fontSize": 10,
                    "marginBottom": 16,
                  }
                }
              >
                error in prop 0 with id0
              </Text>
            </View>
            <View
              style={
                Object {
                  "alignSelf": "stretch",
                  "backgroundColor": "#E3E3E3",
                  "height": 1,
                }
              }
            />
            <View
              style={
                Object {
                  "flexDirection": "column",
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "height": 32,
                    "justifyContent": "space-between",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "fontSize": 12,
                        "fontWeight": "bold",
                      },
                      Object {
                        "color": "#333333",
                      },
                    ]
                  }
                >
                  prop 1 name
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "fontSize": 12,
                      },
                      Object {
                        "color": "#333333",
                      },
                    ]
                  }
                >
                  val 1
                </Text>
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "alignSelf": "stretch",
                "backgroundColor": "#E3E3E3",
                "height": 1,
              }
            }
          />
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
